drone/src/agent/executor.rs

```diff
@@ -37,34 +37,73 @@ impl<T, E: Debug> LogError for Result<T, E> {
     }
 }
 
-#[derive(Clone)]
-pub struct Executor {
-    docker: DockerInterface,
+#[derive(Debug, PartialEq, Eq)]
+enum Signal {
+    /// Tells the executor to interrupt current step to recapture an external status
+    /// change. This signal is sent after an engine detects an external status change,
+    /// e.g. a backend has terminated itself with an error.
+    Interrupt,
```

Comment:
I like this idea, but doesn't it require that the state changes be atomic? for example if the interrupt happens after some non undoable side effect, could we run into incosistent state?

Reply:
That's a good point. For now there's no correctness issue, because interrupts are only fired if a container has died anyway, so the executor immediately discovers that fact and jumps to the appropriate terminal state. But the interrupt itself technically doesn't convey to the executor that the backend is dead, it just nudges the executor to discover that on its own, so `Terminate` doesn't seem appropriate here.

(The reason the signal doesn't convey that the backend has shut down is that we also want to pick up that fact if the drone is resumed and a container has died while the drone was not receiving events from Docker.)